### DISCUSSION TITLE
How to use Combobox in Table filter ?

### DISCUSSION THREAD

**Discussion by mathieu-fournier** [iTwin/iTwinUI#1672] [Q&A] [answered]
I am trying to do a custom filter in the Table. 
It is a text predefined choices multi-selection filter.
I am having trouble using a combobox like stated in the guidelines, the styles does not seem compatible with Table Filter: 
Predefined choices section of the Guidelines : https://ux.bentley.com/itwin/patterns/column-filter.html#type-predefined

I made a codesandbox for it :  https://codesandbox.io/s/table-filter-multi-select-k4r4wz
I tried adding custom styles to make it work, but it does not seems correct.
![image](https://github.com/iTwin/iTwinUI/assets/82905333/b885e97d-1e86-4db7-a25b-8d826bf1c6d3)

At first, I used radio buttons like in the storybook example and it worked just fine : 
https://itwin.github.io/iTwinUI/react/?path=/story/core-table--custom-filter
![image](https://github.com/iTwin/iTwinUI/assets/82905333/e5d7dfc1-bcec-43e0-a74b-149971075364)

Can someone help me using combobox there ?
Or is it okay if I use radio buttons, or should I follow the guidelines. I am not sure what I should do now ?

Note : 
In my case, filtering happens on the server side, so no "frontend" filtering is required here. 
The custom filter only need to send the selection state to the parent component.

**Comment by mayank99** [comment]
I'm not sure I understand what you're trying to achieve here. If you just need a combobox, you should be able to use it inside your custom filter without any custom styles (except maybe a max-width). What difficulty are you facing exactly?

It's also worth noting that those UX guidelines are outdated and unmaintained. We recommend consulting with your product's UX designer for (potentially) new guidelines.

**Reply by mathieu-fournier** [reply]
I forked the codesandbox to have a basic example : 
https://codesandbox.io/s/table-filter-multi-select-base-example-9mrn5l


## The two issues I am facing are : 

I must add,
```
dropdownMenuProps={{
            appendTo: "parent"
          }}
```
, else when clicking on an option, the filter closes and the selection is not registered. 


The styles are overlapping with Filter Buttons and the width changes on selection:
![image](https://github.com/iTwin/iTwinUI/assets/82905333/34933f69-89be-4e04-a5d1-77a22a6255ea)

**Reply by mayank99** [reply]
ah yeah, you need `appendTo: 'parent'` in v2, but it's handled out-of-the-box in v3.

as for the "overlapping with Filter button" part, that should be fine? the dropdown menu can be closed in multiple ways (by clicking on the input _or_ clicking outside it _or_ pressing). the width changing can be handled by using a [responsive fixed width](https://codesandbox.io/s/table-filter-multi-select-base-example-forked-4dnnkc?file=/src/styles.css:197-223).

alternatively, if you don't need search capabilities (e.g. if there aren't a lot of options), you can display a simple list of checkboxes in your custom filter.

does that help?

**Reply by mathieu-fournier** [reply]
It does not help, I still have all those issues.
Try the codesandbox. 
In red:  The buttons are hidden under the dropdown.
In pink: The list should reflect the selected filters. 
![image](https://github.com/iTwin/iTwinUI/assets/82905333/7447a834-ebd3-45d9-b247-2b39bae1e523)



Should we move this discussion to the issue section ? It does not look like I missed something finally.

**Reply by mayank99** [reply]
Could you be more clear about what it is you expect the component to do? because i still do not see an issue that we could handle on our end.

> In red: The buttons are hidden under the dropdown.

I addressed this part above, where i said the dropdown can be closed in multiple ways to show the "Filter" button again.

> In pink: The list should reflect the selected filters.

This part is new (re-reading the entire thread, i don't see any previous mention of it). if you want the filters to update as soon as they are selected in the combobox, then you will need to update your filter state inside `onChange`. If you do this, you can also get rid of the button bar altogether.

see sandbox: https://codesandbox.io/s/table-filter-multi-select-base-example-forked-4dnnkc?file=/src/custom-filter.tsx:1270-1297

**Reply by mathieu-fournier** [reply]
It's the combination of the two that makes the combobox unusable. 

What I expect is : 
- To have the dropdown displayed inside the filter popup. Between the search and the buttons. Not fixed on top. 
- To be able to select many options without it closing. 
- The selection to persist when closing re-opening the filter. 
- To be able to click the filter/clear buttons at all times. 

> I addressed this part above, where i said the dropdown can be closed in multiple ways to show the "Filter" button again.

The dropdown need to stay open because the user might want to select many options. 
The buttons need to be visible at all time, else the user might get confused why selecting an option did not trigger any immediate effect. 

> This part is new (re-reading the entire thread, i don't see any previous mention of it).

This is the first issue I mentionned, but in different words : 
![image](https://github.com/iTwin/iTwinUI/assets/82905333/690eeb7f-35f6-47a2-930a-0492965ba419)
I only want the filter to apply when the Filter button is clicked. In the codesandbox, when a option is selected, the filter closes and the selection is lost (because we weren't able to click the filter button before it closed).

**Reply by r100-stack** [reply]
> What I expect is :
> * The selection to persist when closing re-opening the filter.

To persist the selection, I believe the dependency array on L27 of `App.tsx` needs to contain `Filter` ([updated sandbox](https://codesandbox.io/s/table-filter-multi-select-base-example-forked-9x3jfc?file=/src/App.tsx:663-675))

**Reply by mathieu-fournier** [reply]
I can't make it work with itwin-ui v2 : https://codesandbox.io/s/table-filter-multi-select-base-example-9mrn5l 
Also, the clear buttons would be really nice to have still. 
Plus the possibility to only update the parent when we are done adding all options. Like the 'Filter' button. Else we might trigger a http call for every option clicked.

**Reply by mayank99** [reply, marked as answer]
here is a new sandbox showing how you can avoid `ComboBox` and use `SearchBox` to filter a list of checkboxes while keeping the "Filter"/"Clear" buttons: https://codesandbox.io/s/table-filter-multi-select-base-example-forked-46jpmt?file=/src/custom-filter.tsx

**Reply by mathieu-fournier** [reply]
@mayank99 

The filter never gets the "active state" when a filter is applied. 
![image](https://github.com/iTwin/iTwinUI/assets/82905333/b3c93388-3060-43eb-ad7c-b61de52651c6)

Expected : 
![image](https://github.com/iTwin/iTwinUI/assets/82905333/92aa3c9a-18a9-4764-9fad-6e9f29d6bb02)

**Reply by mayank99** [reply]
well this is a contrived example where the filters are not actually hooked up to the table.

you can use `setFilter` and `clearFilter` props to communicate with the table.

```diff
- Filter: ({ close }) => (
+ Filter: ({ close, setFilter, clearFilter }) => (
   <CustomFilter
     initialSelection={appliedFilters}
     onChange={(filters) => {
       setAppliedFilters(filters);
+      if (filters.length) {
+        setFilter(filters);
+      } else {
+        clearFilter();
+      }
       close();
     }}
     placeholder="placholder"
   />
 )
```

**Reply by mathieu-fournier** [reply]
This only works for frontend filtering, with server-side filtering, when new data is applied to the table, the table filters are reset. 
They are still checked in the Custom filter popup tho. 
Is there a way to prevent the table from reseting it's filter when new data is applied ?

**Reply by mayank99** [reply]
you can pass `autoResetFilters={false}` to the Table
